smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java

```diff
@@ -607,13 +607,39 @@ private void parseMixins(ShapeId id) {
 
         ws();
         do {
+            clearPendingDocs();
             String target = ParserUtils.parseShapeId(this);
             modelFile.addForwardReference(target, resolved -> modelFile.addPendingMixin(id, resolved));
             ws();
-        } while (peek() != '{');
+        } while (!peekEndWith());
+    }
+
+    private boolean peekEndWith() {
+        if (peek() == '{' || peek() == '@' || eof()) {
+            return true;
+        }
+
+        // We could make this more efficient with a prefix trie
+        for (String shapeType : SHAPE_TYPES) {
```

Comment:
Does this mean that shape types basically are keywords in the grammar? I couldn’t name a shape integer for example and try to mix it in…. But only after the first mixin?

Reply:
> Does this mean that shape types basically are keywords in the grammar?

Yes. I think we'd need an explicit continuation marker to make this not the case, that or an explicit end marker.

> I think it also precludes shapes that start with the name of a shape type

This is a bug that can be fixed

Reply:
With the move to bracketing, this isn't relevant anymore